hermes/bin/src/runtime_extensions/hermes/cron/event.rs

```diff
@@ -18,12 +26,207 @@ impl HermesEventPayload for OnCronEvent {
             &self.tag,
             self.last,
         )?;
+        // if the response is `false`, check if the event would
+        // re-trigger, if so, remove it.
+        if !res && !self.last {
+            let app_name = module.store.data().app_name();
+            cron_queue_rm(&app_name.0, self.tag.clone());
+        }
         Ok(())
     }
 }
+
+impl OnCronEvent {
+    /// Get the next scheduled cron event after the optional start timestamp, or after the
+    /// current timestamp.
+    ///
+    /// # Parameters
+    ///
+    /// * `start: Option<CronTimestamp>` - The optional start timestamp. If `None`, the
+    ///   current time is used.
+    ///
+    /// # Returns
+    ///
+    /// * `Some(CronTimestamp)` - The next timestamp for the `OnCronEvent`.
+    /// * `None` if the timestamp could not be calculated.
+    pub(crate) fn tick_after(&self, start: Option<CronTimestamp>) -> Option<CronTimestamp> {
+        let cron = self.cron()?;
+        if cron.any() {
+            let datetime = Self::start_datetime(start)?;
+            let cdt = cron.iter_after(datetime).next()?;
+            let timestamp = cdt.timestamp_nanos_opt()?;
+            timestamp.try_into().ok()
+        } else {
+            None
+        }
+    }
+
+    /// Get the next scheduled cron event from the optional start timestamp, or from the
+    /// current timestamp.
+    ///
+    /// # Parameters
+    ///
+    /// * `start: Option<CronTimestamp>` - The optional start timestamp. If `None`, the
+    ///   current time is used.
+    ///
+    /// # Returns
+    ///
+    /// * `Some(CronTimestamp)` - The next timestamp for the `OnCronEvent`.
+    /// * `None` if the timestamp could not be calculated.
+    pub(crate) fn tick_from(&self, start: Option<CronTimestamp>) -> Option<CronTimestamp> {
+        let cron = self.cron()?;
+        if cron.any() {
+            let datetime = Self::start_datetime(start)?;
+            let cdt = cron.iter_from(datetime).next()?;
+            let timestamp = cdt.timestamp_nanos_opt()?;
+            timestamp.try_into().ok()
+        } else {
+            None
+        }
+    }
+
+    /// Get the w`Cron` from the inner `CronSchedule`.
+    fn cron(&self) -> Option<Cron> {
+        let when = &self.tag.when;
+        when.parse::<Cron>().ok()
+    }
+
+    /// Get the UTC datetime from an optional start timestamp.
+    ///
+    /// Use the `start` timestamp if provided, otherwise use the current time.
+    ///
+    /// Returns `None` if the datetime could not be calculated.:w
+    fn start_datetime(start: Option<CronTimestamp>) -> Option<chrono::DateTime<Utc>> {
+        let datetime = match start {
+            None => Utc::now(),
+            Some(dt) => chrono::DateTime::from_timestamp_nanos(dt.try_into().ok()?),
+        };
+        Some(datetime)
+    }
+}
+
+impl PartialEq for CronTagged {
+    fn eq(&self, other: &Self) -> bool {
+        self.tag == other.tag && self.when == other.when
+    }
+}
+
+impl Eq for CronTagged {}
```

Comment:
Is it possible to use `derive` macro for this ?

Reply:
CronTagged is defined in a WIT file within the project, I don't think that it's possible to use the `derive` macro.